// vuepress-theme-hope/lib/node/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineSidebarObjectConfig = exports.defineSidebarArrayConfig = exports.defineSidebarConfig = exports.defineNavbarConfig = exports.defineThemeConfig = exports.defineHopeConfig = void 0;
const cli_1 = require("@vuepress/cli");
const defineHopeConfig = (config) => (0, cli_1.defineUserConfig)({ theme: "hope", ...config });
exports.defineHopeConfig = defineHopeConfig;
const defineThemeConfig = (themeConfig) => themeConfig;
exports.defineThemeConfig = defineThemeConfig;
const defineNavbarConfig = (navbarConfig) => navbarConfig;
exports.defineNavbarConfig = defineNavbarConfig;
const defineSidebarConfig = (sidebarConfig) => sidebarConfig;
exports.defineSidebarConfig = defineSidebarConfig;
const defineSidebarArrayConfig = (sidebarConfig) => sidebarConfig;
exports.defineSidebarArrayConfig = defineSidebarArrayConfig;
const defineSidebarObjectConfig = (sidebarConfig) => sidebarConfig;
exports.defineSidebarObjectConfig = defineSidebarObjectConfig;
