// vuepress-theme-hope/lib/client/components/CommonWrapper.js
import throttle from "lodash.throttle";
import { Transition, computed, defineComponent, h, onMounted, onUnmounted, ref, resolveComponent, watch, } from "vue";
import { useRouter } from "vue-router";
import { usePageData, usePageFrontmatter } from "@vuepress/client";
import PageFooter from "@theme-hope/components/PageFooter";
import PasswordModal from "@theme-hope/module/encrypt/components/PasswordModal";
import { useMobile, useThemeLocaleData } from "@theme-hope/composables";
import { useGlobalEcrypt } from "@theme-hope/module/encrypt/composables";
import { useSidebarItems } from "@theme-hope/module/sidebar/composables";
import "../styles/common.scss";
export default defineComponent({
    name: "CommonWrapper",
    props: {
        navbar: {
            type: Boolean,
            default: true,
        },
        sidebar: {
            type: Boolean,
            default: true,
        },
    },
    setup(props, { slots }) {
        const router = useRouter();
        const page = usePageData();
        const frontmatter = usePageFrontmatter();
        const themeLocale = useThemeLocaleData();
        const { isGlobalEncrypted, validateGlobalToken } = useGlobalEcrypt();
        const isMobile = useMobile();
        // navbar
        const hideNavbar = ref(false);
        const enableNavbar = computed(() => {
            if (props.navbar === false)
                return false;
            if (frontmatter.value.navbar === false ||
                themeLocale.value.navbar == false)
                return false;
            return Boolean(page.value.title ||
                themeLocale.value.logo ||
                themeLocale.value.repo ||
                themeLocale.value.navbar);
        });
        // sidebar
        const sidebarItems = useSidebarItems();
        const enableSidebar = computed(() => {
            if (props.sidebar === false)
                return false;
            return (frontmatter.value.sidebar !== false &&
                sidebarItems.value.length !== 0 &&
                !frontmatter.value.home);
        });
        const isMobileSidebarOpen = ref(false);
        const isDesktopSidebarOpen = ref(true);
        const toggleMobileSidebar = (value) => {
            isMobileSidebarOpen.value =
                typeof value === "boolean" ? value : !isMobileSidebarOpen.value;
        };
        const toggleDesktopSidebar = (value) => {
            isDesktopSidebarOpen.value =
                typeof value === "boolean" ? value : !isDesktopSidebarOpen.value;
        };
        const touchStart = { x: 0, y: 0 };
        const onTouchStart = (e) => {
            touchStart.x = e.changedTouches[0].clientX;
            touchStart.y = e.changedTouches[0].clientY;
        };
        const onTouchEnd = (e) => {
            const dx = e.changedTouches[0].clientX - touchStart.x;
            const dy = e.changedTouches[0].clientY - touchStart.y;
            if (
            // horizontal swipe
            Math.abs(dx) > Math.abs(dy) * 1.5 &&
                Math.abs(dx) > 40) {
                if (dx > 0 && touchStart.x <= 80)
                    toggleMobileSidebar(true);
                else
                    toggleMobileSidebar(false);
            }
        };
        const enableToc = computed(() => frontmatter.value.toc ||
            (themeLocale.value.toc !== false && frontmatter.value.toc !== false));
        /** Get scroll distance */
        const getScrollTop = () => window.pageYOffset ||
            document.documentElement.scrollTop ||
            document.body.scrollTop ||
            0;
        // close sidebar after navigation
        let unregisterRouterHook;
        let lastDistance = 0;
        const scrollHandler = throttle(() => {
            const distance = getScrollTop();
            // scroll down
            if (lastDistance < distance && distance > 58) {
                if (!isMobileSidebarOpen.value)
                    hideNavbar.value = true;
            }
            // scroll up
            else
                hideNavbar.value = false;
            lastDistance = distance;
        }, 300);
        watch(isMobile, (value) => {
            if (!value)
                toggleMobileSidebar(false);
        });
        onMounted(() => {
            unregisterRouterHook = router.afterEach(() => {
                toggleMobileSidebar(false);
            });
            window.addEventListener("scroll", scrollHandler);
        });
        onUnmounted(() => {
            unregisterRouterHook();
            window.removeEventListener("scroll", scrollHandler);
        });
        return () => {
            var _a;
            return h("div", {
                class: [
                    "theme-container",
                    // classes
                    {
                        "no-navbar": !enableNavbar.value,
                        "no-sidebar": !enableSidebar.value &&
                            !(slots.sidebar || slots.sidebarTop || slots.sidebarBottom),
                        "has-toc": enableToc.value,
                        "hide-navbar": hideNavbar.value,
                        "sidebar-open": isMobile.value
                            ? isMobileSidebarOpen.value
                            : isDesktopSidebarOpen.value,
                    },
                    frontmatter.value.containerClass || "",
                ],
                onTouchStart,
                onTouchEnd,
            }, isGlobalEncrypted.value
                ? h(PasswordModal, { full: true, onVerify: validateGlobalToken })
                : [
                    // navbar
                    enableNavbar.value
                        ? h(resolveComponent("Navbar"), { onToggleSidebar: () => toggleMobileSidebar() }, {
                            left: () => { var _a; return (_a = slots.navbarLeft) === null || _a === void 0 ? void 0 : _a.call(slots); },
                            center: () => { var _a; return (_a = slots.navbarCenter) === null || _a === void 0 ? void 0 : _a.call(slots); },
                            right: () => { var _a; return (_a = slots.navbarRight) === null || _a === void 0 ? void 0 : _a.call(slots); },
                            screenTop: () => { var _a; return (_a = slots.navScreenTop) === null || _a === void 0 ? void 0 : _a.call(slots); },
                            screenBottom: () => { var _a; return (_a = slots.navScreenBottom) === null || _a === void 0 ? void 0 : _a.call(slots); },
                        })
                        : null,
                    // sidebar mask
                    h(Transition, { name: "fade" }, () => isMobileSidebarOpen.value
                        ? h("div", {
                            class: "sidebar-mask",
                            onClick: () => toggleMobileSidebar(false),
                        })
                        : null),
                    // toggle sidebar button
                    h(Transition, { name: "fade" }, () => isMobile.value
                        ? null
                        : h("div", {
                            class: "toggle-sidebar-wrapper",
                            onClick: () => toggleDesktopSidebar(),
                        }, h("span", {
                            class: [
                                "arrow",
                                isDesktopSidebarOpen.value ? "left" : "right",
                            ],
                        }))),
                    // sidebar
                    h(resolveComponent("Sidebar"), {}, {
                        ...(slots.sidebar
                            ? { default: () => { var _a; return (_a = slots.sidebar) === null || _a === void 0 ? void 0 : _a.call(slots); } }
                            : {}),
                        top: () => { var _a; return (_a = slots.sidebarTop) === null || _a === void 0 ? void 0 : _a.call(slots); },
                        bottom: () => { var _a; return (_a = slots.sidebarBottom) === null || _a === void 0 ? void 0 : _a.call(slots); },
                    }),
                    (_a = slots.default) === null || _a === void 0 ? void 0 : _a.call(slots),
                    h(PageFooter),
                ]);
        };
    },
});
