// vuepress-theme-hope/lib/client/module/navbar/components/LanguageDropdown.js
import { defineComponent, h } from "vue";
import { I18nIcon } from "@theme-hope/module/navbar/components/icons";
import DropdownLink from "@theme-hope/module/navbar/components/DropdownLink";
import { useNavbarLanguageDropdown } from "@theme-hope/module/navbar/composables";
export default defineComponent({
    name: "LanguageDropdown",
    setup() {
        const dropdown = useNavbarLanguageDropdown();
        return () => dropdown.value
            ? h(DropdownLink, { class: "i18n-dropdown", config: dropdown.value }, {
                title: () => {
                    var _a;
                    return h(I18nIcon, {
                        ariaLabel: (_a = dropdown.value) === null || _a === void 0 ? void 0 : _a.ariaLabel,
                        style: {
                            width: "1rem",
                            height: "1rem",
                            verticalAlign: "middle",
                        },
                    });
                },
            })
            : null;
    },
});
