// vuepress-theme-hope/lib/client/module/navbar/components/NavActions.js
import { defineComponent, h, resolveComponent } from "vue";
import LanguageDropdown from "@theme-hope/module/navbar/components/LanguageDropdown";
import RepoLink from "@theme-hope/module/navbar/components/RepoLink";
import OutlookButton from "@theme-hope/module/outlook/components/OutlookButton";
import ToggleNavbarButton from "@theme-hope/module/navbar/components/ToggleNavbarButton";
import "../styles/navbar-actions.scss";
export default defineComponent({
    name: "NavActions",
    props: {
        showScreen: { type: Boolean, default: false },
    },
    emits: ["toggleScreen"],
    setup(props, { emit, slots }) {
        return () => {
            var _a, _b;
            return h("div", { class: "nav-actions-wrapper" }, [
                (_a = slots.before) === null || _a === void 0 ? void 0 : _a.call(slots),
                h("div", { class: "nav-item" }, h(LanguageDropdown)),
                h(RepoLink),
                h(OutlookButton),
                h(resolveComponent("NavbarSearch")),
                h(ToggleNavbarButton, {
                    active: props.showScreen,
                    onToggle: () => emit("toggleScreen"),
                }),
                (_b = slots.after) === null || _b === void 0 ? void 0 : _b.call(slots),
            ]);
        };
    },
});
